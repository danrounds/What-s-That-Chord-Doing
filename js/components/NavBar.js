import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router';
import MediaQuery from 'react-responsive';

import * as actions from '../actions';

import KeyboardShortcutsOnOff from './KeyboardShortcutsOnOff';

export class NavBar extends React.Component {
    constructor(props) {
        super(props);
        this.logOff = this.logOff.bind(this);
    }
    
    logOff() {
        this.props.dispatch(actions.logOff());
    }

    render() {
        const loggedIn = Boolean(this.props.api.authToken);
        return (
            <div className='nav-bar'>
              <div className={this.props.parent === 'Home'
                   ? "nav-button nav-button-selected" : "nav-button"}>
                <Link className="nav-text" to="/">Exercises</Link>
              </div>

              <div className={this.props.parent === 'WhatIs'
                   ? "nav-button nav-button-selected" : "nav-button"}>
                <Link className="nav-text" to="/what-is-this">What is this?</Link>
              </div>

              <MediaQuery minDeviceWidth={800}>
                {(matches) => {
                    let showHints;
                    if (this.props.displayKeyboardShortcuts === undefined || this.props.displayKeyboardShortcuts === null)
                        showHints = matches;
                    else
                        showHints = this.props.displayKeyboardShortcuts;
                    return (<KeyboardShortcutsOnOff showButton={matches} showHints={showHints}/>);
                }}
              </MediaQuery>

              <div className={this.props.parent === 'Scores'
                   ? "nav-button nav-button-selected" : "nav-button"}>
                <Link className="nav-text" to="/scores">Scores</Link>
              </div>

              <div className={this.props.parent === 'LogIn' ? "nav-button nav-button-selected" : "nav-button"}>
                {loggedIn ? <a className="nav-text" href="javascript:void(0)" onClick={this.logOff}>Log off</a>
                   : <Link className="nav-text" to="/log-in-or-register">Log in</Link>}
              </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    displayKeyboardShortcuts: state.ui.displayKeyboardShortcuts,
    api: state.api,
});

export default connect(mapStateToProps)(NavBar);
