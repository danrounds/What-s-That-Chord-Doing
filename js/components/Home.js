import React from 'react';
import {Link} from 'react-router';
import {Provider} from 'react-redux';

import store from '../store';

import NavBar from './NavBar';
import LessonsNavigator from './LessonsNavigator';
import KeyboardNavLessonListOverlay from './KeyboardNavLessonListOverlay';

export default class Home extends React.Component {
    constructor(props) {
        super(props);
        this.Home = (
            <div>
              <NavBar parent="Home"/>
              <LessonsNavigator />
              <KeyboardNavLessonListOverlay />
            </div>
        );
    }

    componentDidMount() {
        this.props.instrument.then(piano => {
            piano.stop();
        });
    }


    render(props) {
        return(
              <Provider store={store}>
                {this.Home}
              </Provider>
        );
    }
}
