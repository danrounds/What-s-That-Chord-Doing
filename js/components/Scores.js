import React from 'react';
import {connect} from 'react-redux';

import HighScoresDisplay from './HighScoresDisplay';
import MyScoresDisplay from './MyScoresDisplay';
import NavBar from './NavBar';
import * as actions from '../actions';

export class Scores extends React.Component {
    constructor() {
        super();
        this.state = { showScores: 'highScores' };
        this.setColor = this.setColor.bind(this);
        this.showHighScores = this.showHighScores.bind(this);
        this.showMyScores = this.showMyScores.bind(this);
    }

    setColor(selected) {
        if (selected === this.state.showScores)
            return { backgroundColor: 'pink' };
        return {};
    }

    showHighScores() {
        this.setState({ showScores: 'highScores' });
    }

    showMyScores() {
        this.setState({ showScores: 'myScores' });
    }

    componentDidMount() {
        this.props.instrument.then(piano => piano.stop());
    }

    render() {
        return (
            <div>
              <NavBar parent="Scores"/>

              <button onClick={this.showHighScores} style={this.setColor('highScores')}>
                High Scores
              </button>

              <button onClick={this.showMyScores} style={this.setColor('myScores')}>
                My Scores
              </button>

              {this.state.showScores === 'highScores' ? <HighScoresDisplay /> : null}
              {this.state.showScores === 'myScores' ? <MyScoresDisplay /> : null }

            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    gameType: state.game.gameType || 'allChords',
    api: state.api,
});

export default connect(mapStateToProps)(Scores);
