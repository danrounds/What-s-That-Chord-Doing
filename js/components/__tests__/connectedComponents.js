import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Game } from '../Game';
import Home from '../Home';
import { NavBar } from '../NavBar';
import Scores from '../Scores';
import { WhatIsThis } from '../WhatIsThis';

// These are components that really, /really/ want a Store passed to them,
// unless we use the shallow renderer. These are all rendering-oriented tests

describe('Score-based components', () => {

    test('Game', () => {
        const dispatch = (arg) => null;
        const comp = shallow(<Game dispatch={dispatch} mode="allChords" inversions={false} api={{}}/>);
        expect(toJson(comp)).toMatchSnapshot();
    });

    test('Home', () => {
        const comp = shallow(<Home instrument={{ then: (arg) => null }} />);
        expect(toJson(comp)).toMatchSnapshot();
    });

    test('NavBar', () => {
        const comp = shallow(<NavBar parent="Home" api={{ authToken: 'some_string' }}/>);
        expect(toJson(comp)).toMatchSnapshot();
    });

    test('Scores', () => {
        const comp = shallow(<Scores />);
        expect(toJson(comp)).toMatchSnapshot();
    });

    test('WhatIsThis', () => {
        const comp = shallow(<WhatIsThis introChordSequence={[]} />);
        expect(toJson(comp)).toMatchSnapshot();
    });    

});
