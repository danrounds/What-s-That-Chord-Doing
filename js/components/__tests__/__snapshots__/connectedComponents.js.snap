// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Score-based components Game 1`] = `
<div
  onBlur={[Function]}
  onKeyDown={[Function]}
  tabIndex="0"
>
  <Connect(NavBar) />
  <div
    className="game"
  >
    <Connect(GameModeText) />
    <Connect(KofN) />
    <Connect(Staves) />
    <Connect(Status)
      inversions={false}
      mode="allChords"
    />
    <Connect(AnswerEntry) />
    <Connect(PlayAudio) />
    <Connect(StartNewGame) />
    <Connect(GiveUpOrNext) />
    <Connect(SubmitMyScore) />
  </div>
</div>
`;

exports[`Score-based components Home 1`] = `
<div>
  <Connect(NavBar)
    parent="Home"
  />
  <Connect(LessonsNavigator) />
  <Connect(KeyboardNavLessonListOverlay) />
</div>
`;

exports[`Score-based components NavBar 1`] = `
<div
  className="nav-bar"
>
  <div
    className="nav-button nav-button-selected"
  >
    <Link
      className="nav-text"
      onlyActiveOnIndex={false}
      style={Object {}}
      to="/"
    >
      Exercises
    </Link>
  </div>
  <div
    className="nav-button"
  >
    <Link
      className="nav-text"
      onlyActiveOnIndex={false}
      style={Object {}}
      to="/what-is-this"
    >
      What is this?
    </Link>
  </div>
  <MediaQuery
    minDeviceWidth={800}
    values={Object {}}
  />
  <div
    className="nav-button"
  >
    <Link
      className="nav-text"
      onlyActiveOnIndex={false}
      style={Object {}}
      to="/scores"
    >
      Scores
    </Link>
  </div>
  <div
    className="log-in-btn nav-button"
  >
    <a
      className="nav-text"
      href="javascript:void(0)"
      onClick={[Function]}
    >
      Log off
    </a>
  </div>
</div>
`;

exports[`Score-based components Scores 1`] = `
<div>
  <Connect(NavBar)
    parent="Scores"
  />
  <div
    className="huge-btn-div"
  >
    <button
      className="huge-btn huge-btn-active"
      onClick={[Function]}
    >
      High Scores
    </button>
    <button
      className="huge-btn"
      onClick={[Function]}
    >
      My Scores
    </button>
  </div>
  <Connect(HighScoresDisplay) />
</div>
`;

exports[`Score-based components WhatIsThis 1`] = `
<div
  tabIndex="1"
>
  <Connect(NavBar)
    parent="WhatIs"
  />
  <div
    className="what-is-this"
  >
    <div
      className="what-is-it-section"
    >
      <h1
        className="what-is-this-h"
      >
        What does it do?
      </h1>
      <p>
        If you've played music, you've heard of the 
        <em>
          I-IV-V
        </em>
         chord progression.
      </p>
      <p>
        Or maybe you've 
        <a
          className="what-is-link"
          href="https://www.youtube.com/watch?v=oOlDewpCfZQ"
          rel="noopener noreferrer"
          target="_blank"
        >
          seen the video on the ubiquitous I-IV-vi-V chord progression.
        </a>
      </p>
      <p>
        This app teaches you to hear that 
        <em>
          I-IV-V
        </em>
        ... or 
        <em>
          vii°
        </em>
        ... 
        <em>
          ♭II
        </em>
         ...
      </p>
    </div>
    <div
      className="what-is-it-section"
    >
      <h1
        className="what-is-this-h"
      >
        How does it work?
      </h1>
      <p>
        This explains the basic mechanics of the game—what you'll hear and why.
      </p>
      <p>
        <em>
          If you're not sure what a "function ear trainer" is, check out the 
          <span
            className="italic"
          >
            Why This Game
          </span>
           section, below.
        </em>
      </p>
      <p>
        <em>
          First,
        </em>
         you'll hear a four chord intro (
        <em>
          I-IV-V-I
        </em>
        ), which will unambiguously establish a key:
      </p>
      <div
        className="btn-container"
      >
        <button
          className="what-is-this-btn"
          onClick={[Function]}
        >
          Press to hear the intro
        </button>
      </div>
      <p>
        <em>
          After a brief pause, you'll hear a fifth chord:
        </em>
         the actual, prompted question that you're to answer:
      </p>
      <div
        className="btn-container"
      >
        <button
          className="what-is-this-btn"
          onClick={[Function]}
        >
          Press to hear the prompt
        </button>
      </div>
      <p>
        <em>
          This is what it'll actually sound like, in-game:
        </em>
      </p>
      <div
        className="btn-container"
      >
        <button
          className="what-is-this-btn"
          onClick={[Function]}
        >
          Press to hear the intro and prompt
        </button>
      </div>
      <p>
        <em>
          Now, you pick the roman numeral of the chord you've just heard.
        </em>
         It'll be hard, at first!
      </p>
    </div>
    <div
      className="what-is-it-section"
      id="why"
    >
      <h1
        className="what-is-this-h"
      >
        Why this game?
      </h1>
      <p>
        <a
          className="what-is-link"
          href="http://tonedear.com/ear-training/chord-identification"
          rel="noopener noreferrer"
          target="_blank"
        >
          Conventional chord training apps
        </a>
         are useful, but they have a problem: 
        <em>
          They ignore musical context.
        </em>
      </p>
      <p>
        They have you identify chords based on their "chord quality"—
        <em>
          major, minor, diminished, add6, etc.
        </em>
      </p>
      <p>
        Again, 
        <span
          className="italic"
        >
          this is useful
        </span>
        , but we're ignoring part of how people actually hear music.
      </p>
      <p>
        When we listen to music, we automatically relate chords and notes to the context of a key (whether we know what a key is or not). A C major chord in the key of C Major sounds very different than a C Major chord in F# minor.
      </p>
      <p>
        <a
          className="what-is-link"
          href="http://www.harding.edu/gclayton/color/images/simultcontr/ch03%20perceiving%20color162.png"
          rel="noopener noreferrer"
          target="_blank"
        >
          Something very similar happens in the way we perceive color.
        </a>
      </p>
      <p>
        The 
        <a
          className="what-is-link"
          href="https://www.youtube.com/watch?v=WZ6qEomjWM4"
          rel="noopener noreferrer"
          target="_blank"
        >
          I-IV-vi-V chord progression sounds familiar
        </a>
         precisely because of contextual hearing.
      </p>
      <p>
        Even though songs that use this progression aren't always (or even usually) in the same key, we still hear them as "the same," in some sense, 
        <span
          className="italic"
        >
          precisely
        </span>
         because our hearing is contextual.
      </p>
    </div>
  </div>
</div>
`;
